File: hu/codes/java/chapter_computational_complexity/worst_best_time_complexity.java
/**
 * File: worst_best_time_complexity.java
 * Created Time: 2022-11-25
 * Author: krahets (krahets@163.com)
 */

package chapter_computational_complexity;

import java.util.*;

public class worst_best_time_complexity {
    /* Tömb generálása { 1, 2, ..., n } elemekkel, véletlenszerű sorrendben */
    static int[] randomNumbers(int n) {
        Integer[] nums = new Integer[n];
        // nums = { 1, 2, 3, ..., n } tömb generálása
        for (int i = 0; i < n; i++) {
            nums[i] = i + 1;
        }
        // Tömbelemek véletlenszerű összekeverése
        Collections.shuffle(Arrays.asList(nums));
        // Integer[] -> int[]
        int[] res = new int[n];
        for (int i = 0; i < n; i++) {
            res[i] = nums[i];
        }
        return res;
    }

    /* Az 1-es szám indexének megkeresése a nums tömbben */
    static int findOne(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            // Ha az 1-es elem a tömb elején van, legjobb eset időbonyolultsága O(1)
            // Ha az 1-es elem a tömb végén van, legrosszabb eset időbonyolultsága O(n)
            if (nums[i] == 1)
                return i;
        }
        return -1;
    }

    /* Főprogram */
    public static void main(String[] args) {
        for (int i = 0; i < 10; i++) {
            int n = 100;
            int[] nums = randomNumbers(n);
            int index = findOne(nums);
            System.out.println("\nArray [ 1, 2, ..., n ] after shuffling = " + Arrays.toString(nums));
            System.out.println("Index of number 1 is " + index);
        }
    }
}

File: hu/codes/dart/chapter_computational_complexity/worst_best_time_complexity.dart
/**
 * File: worst_best_time_complexity.dart
 * Created Time: 2023-02-12
 * Author: Jefferson (JeffersonHuang77@gmail.com)
 */

/* { 1, 2, ..., n } elemeket tartalmazó tömb generálása, véletlenszerű sorrendben */
List<int> randomNumbers(int n) {
  final nums = List.filled(n, 0);
  // nums = { 1, 2, 3, ..., n } tömb generálása
  for (var i = 0; i < n; i++) {
    nums[i] = i + 1;
  }
  // Tömb elemeinek véletlenszerű keverése
  nums.shuffle();

  return nums;
}

/* Az 1-es szám indexének megkeresése a nums tömbben */
int findOne(List<int> nums) {
  for (var i = 0; i < nums.length; i++) {
    // Ha az 1-es elem a tömb elején van, legjobb esetbeli időbonyolultság O(1) érhető el
    // Ha az 1-es elem a tömb végén van, legrosszabb esetbeli időbonyolultság O(n) érhető el
    if (nums[i] == 1) return i;
  }

  return -1;
}

/* Főprogram */
void main() {
  for (var i = 0; i < 10; i++) {
    int n = 100;
    final nums = randomNumbers(n);
    int index = findOne(nums);
    print('\nArray [ 1, 2, ..., n ] after shuffling = $nums');
    print('Index of number 1 is + $index');
  }
}
